File: 22-generate-parentheses/generate-parentheses.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<String> result = new ArrayList<>();

    public boolean isValid(String s) {
        int count = 0;
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (ch == '(')
                count++;
            else
                count--;
            if (count < 0)
                return false;
        }
        return count == 0;
    }

    public void backTrack(StringBuilder curr, int n) {
        if (curr.length() == 2 * n) {
            if (isValid(curr.toString())) {
                result.add(curr.toString());
            }
            return;
        }
        curr.append('(');
        backTrack(curr, n);
        curr.deleteCharAt(curr.length() - 1);

        curr.append(')');
        backTrack(curr, n);
        curr.deleteCharAt(curr.length() - 1);
    }

    public List<String> generateParenthesis(int n) {
        StringBuilder curr = new StringBuilder();
        backTrack(curr, n);
        return result;
    }
}

File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:

    vector <string> result;

    bool isValid(string s){
        
        int count = 0;
        for(char &ch: s){
            if(ch == '(')
            count++;
            else
            count--;
            if(count<0)
            return false;
        }
        return count==0; 
    }

    void backTrack( string &curr , int n ){
        if(curr.length() == 2*n){
            if(isValid(curr)){
                result.push_back(curr);
            }
            return ;
        }
        curr.push_back('(');
        backTrack(curr,n);
        curr.pop_back();

        curr.push_back(')');
        backTrack(curr ,n );
        curr.pop_back();
    }

    vector<string> generateParenthesis(int n) {
        string curr = "";

        backTrack(curr,n);

        return result;
    }
};